cmd/gpu_nfdhook/labeler.go

```diff
@@ -171,33 +171,69 @@ func (l *labeler) createCapabilityLabels(cardNum string, numTiles uint64) {
 	}
 	defer file.Close()
 
-	// define strings to search from the file, and the actions to take in order to create labels from those strings (as funcs)
+	gen := ""
+	media := ""
+	graphics := ""
+	// define string prefixes to search from the file, and the actions to take in order to create labels from those strings (as funcs)
 	searchStringActionMap := map[string]func(string){
-		"platform: ": func(platformName string) {
+		"platform:": func(platformName string) {
 			l.labels.addNumericLabel(labelNamespace+"platform_"+platformName+".count", 1)
 			l.labels[labelNamespace+"platform_"+platformName+".tiles"] = strconv.FormatInt(int64(numTiles), 10)
 			l.labels[labelNamespace+"platform_"+platformName+".present"] = "true"
 		},
-		"gen: ": func(genName string) {
-			l.labels[labelNamespace+"platform_gen"] = genName
+		// there's also display block version, but that's not relevant
+		"media version:": func(version string) {
+			l.labels[labelNamespace+"media_version"] = version
+			media = version
+		},
+		"graphics version:": func(version string) {
+			l.labels[labelNamespace+"graphics_version"] = version
+			graphics = version
+		},
+		"gen:": func(version string) {
+			l.labels[labelNamespace+"platform_gen"] = version
+			gen = version
 		},
 	}
 
 	// Finally, read the file, and try to find the matches. Perform actions and reduce the search map size as we proceed. Return at 0 size.
 	scanner := bufio.NewScanner(file)
 	for scanner.Scan() {
-		for searchString, action := range searchStringActionMap {
-			var stringValue string
-			n, _ := fmt.Sscanf(scanner.Text(), searchString+"%s", &stringValue)
-			if n > 0 {
-				action(stringValue)
-				delete(searchStringActionMap, searchString)
-				if len(searchStringActionMap) == 0 {
-					return
-				}
-				break
+		for prefix, action := range searchStringActionMap {
+			line := scanner.Text()
+			if !strings.HasPrefix(line, prefix) {
+				continue
+			}
+			fields := strings.Split(line, ": ")
+			if len(fields) == 2 {
+				action(fields[1])
+			} else {
+				klog.Warningf("invalid '%s' line format: '%s'", file.Name(), line)
+			}
+			delete(searchStringActionMap, prefix)
+			if len(searchStringActionMap) == 0 {
+				return
 			}
+			break
+		}
```

Comment:
Do I understand it right that we take a first entry in `searchStringActionMap` (which can be any) and look for it in the first line. If not found we take another entry and look for it in the second line (but not in the first list). If we are lucky and the second entry of the unordered map coincided with the second line then we either stop iterating or return from the function at all.

Reply:
Admittedly the code is a bit convoluted, partly because the original was also, before Eero touched it. My folt.

The code does indeed take the first entry in the searchStringActionMap and looks for the prefix from THE line. If the prefix is NOT found, another string from the map is attempted (the continue -path) with the same line from the scanner. The scanner does not progress forward with a call to Text(). Perhaps one could take the "line := scanner.Text()" out of the loop? Probably.

If the string IS found, well eventually either the whole function returns in case there are no more strings left in the map, or we proceed with the break-path which means the for loop exits, and another line is read from the scanner. That means, another line from the capability file gets investigated with the same reducing search map.